In [195]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from sklearn import neighbors, datasets
from sklearn.metrics import confusion_matrix


from sklearn.neighbors import NearestNeighbors, KNeighborsClassifier
from sklearn.model_selection import train_test_split


data = pd.read_csv("UniversalBank.csv")

print(data.head(20))

    ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  Personal Loan  \
0    1   25           1      49     91107       4    1.6          1         0              0   
1    2   45          19      34     90089       3    1.5          1         0              0   
2    3   39          15      11     94720       1    1.0          1         0              0   
3    4   35           9     100     94112       1    2.7          2         0              0   
4    5   35           8      45     91330       4    1.0          2         0              0   
5    6   37          13      29     92121       4    0.4          2       155              0   
6    7   53          27      72     91711       2    1.5          2         0              0   
7    8   50          24      22     93943       1    0.3          3         0              0   
8    9   35          10      81     90089       3    0.6          2       104              0   
9   10   34           9     180     9302

In [62]:
data.shape

(5000, 14)

In [63]:
data.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [64]:
pd.set_option('display.width', 100)
pd.set_option('precision', 3)
data.describe()

ID       Age  Experience    Income   ZIP Code    Family     CCAvg  Education  \
count  5000.00  5000.000    5000.000  5000.000   5000.000  5000.000  5000.000   5000.000   
mean   2500.50    45.338      20.105    73.774  93152.503     2.396     1.938      1.881   
std    1443.52    11.463      11.468    46.034   2121.852     1.148     1.748      0.840   
min       1.00    23.000      -3.000     8.000   9307.000     1.000     0.000      1.000   
25%    1250.75    35.000      10.000    39.000  91911.000     1.000     0.700      1.000   
50%    2500.50    45.000      20.000    64.000  93437.000     2.000     1.500      2.000   
75%    3750.25    55.000      30.000    98.000  94608.000     3.000     2.500      3.000   
max    5000.00    67.000      43.000   224.000  96651.000     4.000    10.000      3.000   

       Mortgage  Personal Loan  Securities Account  CD Account    Online  CreditCard  
count  5000.000       5000.000            5000.000    5000.000  5000.000    5000.000  
mean     56.499          0.096               0.104       0.060     0.597       0.294  
std     101.714          0.295               0.306       0.238     0.491       0.456  
min       0.000          0.000               0.000       0.000     0.000       0.000  
25%       0.000          0.000               0.000       0.000     0.000       0.000  
50%       0.000          0.000               0.000       0.000     1.000       0.000  
75%     101.000          0.000               0.000       0.000     1.000       1.000  
max     635.000          1.000               1.000       1.000     1.000       1.000

In [141]:
data_dummies = pd.get_dummies(data,['Edu'],'_',False,['Education'])
data_dummies.set_index('ID', inplace = True)

In [158]:
train, test = train_test_split(data_dummies, test_size=0.4, random_state=42, shuffle=True)


train_y = train.filter(items=['Personal Loan'])
train_x = train.drop(['Personal Loan','ZIP Code'], axis=1)

test_y = test.filter(items=['Personal Loan'])
test_x = test.drop(['Personal Loan','ZIP Code'], axis=1)

In [143]:
n_neighbors = 1
nbrs = KNeighborsClassifier(n_neighbors=n_neighbors)
nbrs.fit(train_x,train_y)

# Age = 40, Experience = 10, Income = 84, Family = 2, CCAvg = 2, Mortgage = 0, Securities Account = 0, CD Account = 0, Online = 1, and Credit Card = 1, Education_1 = 0, Education_2 = 1, Education_3 = 0,
customer = np.array([[40, 10, 84, 2, 2, 0, 0, 0, 1, 1, 0,1,0]])

print(train_x.shape)
print(customer.shape)
print(train_x.head(20))


print(nbrs.predict(customer))

(3000, 13)
(1, 13)
      Age  Experience  Income  Family  CCAvg  Mortgage  Securities Account  CD Account  Online  \
ID                                                                                               
4577   55          30      41       2   0.60         0                   0           0       1   
1089   59          35      95       1   3.80         0                   0           0       1   
480    60          36     132       2   6.00       470                   0           0       1   
728    62          37      18       3   1.30         0                   0           0       1   
3784   60          34      51       3   1.40         0                   0           0       0   
977    54          30      24       4   0.20         0                   1           0       0   
3785   30           6     115       4   3.80         0                   0           0       1   
2978   35          10     161       1   4.10         0                   1           0       1   
1

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


The customer is predicted to not opt for the loan

In [144]:
for i in range(1,15):
    nbrs = KNeighborsClassifier(n_neighbors=i)
    nbrs.fit(train_x, train_y)
    print("K="+str(i)+" "+str(nbrs.score(test_x,test_y)))
    

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversion

K=1 0.896
K=2 0.907
K=3 0.901
K=4 0.9085


C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversion

K=5 0.9035
K=6 0.905
K=7 0.903
K=8 0.9045


C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


K=9 0.899
K=10 0.9025
K=11 0.903
K=12 0.907
K=13 0.904
K=14 0.902


C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


K=12 has the highest score

12 neighbors is best because it gives the highest accuracy over the validation set

In [198]:
n_neighbors = 12
nbrs = KNeighborsClassifier(n_neighbors=n_neighbors)
nbrs.fit(train_x,train_y)

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=12, p=2,
                     weights='uniform')

In [199]:
test_values = test_x.to_dict('index')
test_list = []

# print(test_values)

for index, rows in test_y.iterrows():
    test_values[index]['Personal Loan'] = rows['Personal Loan']

pred = []
actual = []
for x in test_values:
    cust = list(test_values[x].values())
    test_list.append(cust)
    pred.append(nbrs.predict([cust[:-1]])[0])
    actual.append(cust[-1:][0])


In [200]:
confusion_matrix(actual,pred)

array([[1771,   28],
       [ 158,   43]], dtype=int64)

This means that
rows are actual columns are predictions

1771 predictions were 0 and actual 0
28   predictions were 1 and acutal 0
158  predictions were 0 and actual 1
43   predictions were 1 and actual 1

This means that the algorithm still highly favors inaccuratly predicting 0 for loan aplicants who would have accepted
False Negative Ratio = .0819

This is a false negative, and represents missed opportunities for the bank

In [201]:
customer = np.array([[40, 10, 84, 2, 2, 1, 0, 0, 1, 1, 0,1,0]])

print(nbrs.predict(customer))

[0]


In [202]:
train, temp = train_test_split(data_dummies, test_size=0.5, random_state=42, shuffle=True)
val, test = train_test_split(temp, test_size=0.4, random_state=42, shuffle=True)

train_y = train.filter(items=['Personal Loan'])
train_x = train.drop(['Personal Loan','ZIP Code'], axis=1)

val_y = val.filter(items=['Personal Loan'])
val_x = val.drop(['Personal Loan','ZIP Code'], axis=1)

test_y = test.filter(items=['Personal Loan'])
test_x = test.drop(['Personal Loan','ZIP Code'], axis=1)

In [204]:
for i in range(1,15):
    nbrs = KNeighborsClassifier(n_neighbors=i)
    nbrs.fit(train_x, train_y)
    print("K="+str(i)+" "+str(nbrs.score(val_x,val_y)))

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversion

K=1 0.882
K=2 0.8946666666666667
K=3 0.8893333333333333
K=4 0.8953333333333333
K=5 0.8906666666666667
K=6 0.892

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversion


K=7 0.89
K=8 0.888
K=9 0.8853333333333333
K=10 0.8906666666666667
K=11 0.8906666666666667
K=12 0.8926666666666667
K=13 0.8886666666666667
K=14 0.8933333333333333


C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversion

K = 14 is the best score

In [205]:
n_neighbors = 14
nbrs = KNeighborsClassifier(n_neighbors=n_neighbors)
nbrs.fit(train_x,train_y)

C:\Users\EricC\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=14, p=2,
                     weights='uniform')

In [206]:
test_values = test_x.to_dict('index')
test_list = []

# print(test_values)

for index, rows in test_y.iterrows():
    test_values[index]['Personal Loan'] = rows['Personal Loan']

pred = []
actual = []
for x in test_values:
    cust = list(test_values[x].values())
    test_list.append(cust)
    pred.append(nbrs.predict([cust[:-1]])[0])
    actual.append(cust[-1:][0])

In [207]:
confusion_matrix(actual,pred)

array([[898,  11],
       [ 77,  14]], dtype=int64)

This means that
rows are actual columns are predictions

898  predictions were 0 and actual 0
11   predictions were 1 and acutal 0
77   predictions were 0 and actual 1
14   predictions were 1 and actual 1

This means that the algorithm still highly favors inaccuratly predicting 0 for loan aplicants who would have accepted
False Negative Ratio = .0790

This is a false negative, and represents missed opportunities for the bank but because the data used a training/validation/test split the false negative ratio went down slightly. This happens because the test data has never been seen by the algorithm and the validation data was used to select the K value for the algorithm. The results are more robust because the data used to select the K value was different from the data used to test the algorithm.